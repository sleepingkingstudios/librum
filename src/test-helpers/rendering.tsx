import * as React from 'react';

import { MemoryRouter } from 'react-router-dom';
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import { ThemeContext } from '@themes/context';
import { defaultTheme } from '@themes/theme';

interface themeProperties {
  [index: string]: string;
}

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true;
  theme?: themeProperties;
}

interface withRouterOptions {
  initialEntries?: Array<string>;
}

interface withThemeOptions {
  theme: themeProperties;
}

export const withRouter = (
  component: React.ReactElement,
  { initialEntries = ['/'] }: withRouterOptions,
): React.ReactElement => (
  <MemoryRouter initialEntries={initialEntries}>
    { component }
  </MemoryRouter>
);

export const withTheme = (
  component: React.ReactElement,
  { theme }: withThemeOptions
): React.ReactElement => (
  <ThemeContext.Provider value={{ ...defaultTheme, ...theme }}>
    { component }
  </ThemeContext.Provider>
);

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions
): RenderResult => {
  if (!options) { return render(component); }

  let wrapped: React.ReactElement = component;

  if ('router' in options && options.router) {
    wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
