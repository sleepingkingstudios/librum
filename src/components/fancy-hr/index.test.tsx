import * as React from 'react';

// import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

import { FancyHr } from './index';
import { render } from '@test-helpers/rendering';

describe('<FancyHr>', () => {
  const theme = {
    hr: '@hrBase border-red-500',
    hrDanger: '@hrBase border-green-500',
  };

  it('should match the snapshot', () => {
    const { asFragment } = render(<FancyHr />, { theme });

    expect(asFragment()).toMatchSnapshot();
  });

  describe('with className: value', () => {
    it('should append the className', () => {
      const className = 'some-class-name';
      const { asFragment } =
        render(<FancyHr className={className} />, { theme });

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with style: value', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(<FancyHr style="hr-danger" />, { theme });

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
