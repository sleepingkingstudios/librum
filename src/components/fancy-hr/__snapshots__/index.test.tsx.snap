// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FancyHr> should match the snapshot 1`] = `
<DocumentFragment>
  <hr
    class="fancy-hr border-red-500"
  />
</DocumentFragment>
`;

exports[`<FancyHr> with className: value should append the className 1`] = `
<DocumentFragment>
  <hr
    class="fancy-hr border-red-500 some-class-name"
  />
</DocumentFragment>
`;

exports[`<FancyHr> with style: value should match the snapshot 1`] = `
<DocumentFragment>
  <hr
    class="fancy-hr border-green-500"
  />
</DocumentFragment>
`;
